editor/import/scene_importer_mesh.cpp

```diff
@@ -436,6 +452,338 @@ Dictionary EditorSceneImporterMesh::_get_data() const {
+Ref<Shape3D> EditorSceneImporterMesh::create_trimesh_shape() const {
+	Vector<Face3> faces = get_faces();
+	if (faces.size() == 0) {
+		return Ref<Shape3D>();
+	}
+
+	Vector<Vector3> face_points;
+	face_points.resize(faces.size() * 3);
+
+	for (int i = 0; i < face_points.size(); i += 3) {
+		Face3 f = faces.get(i / 3);
+		face_points.set(i, f.vertex[0]);
+		face_points.set(i + 1, f.vertex[1]);
+		face_points.set(i + 2, f.vertex[2]);
+	}
+
+	Ref<ConcavePolygonShape3D> shape = memnew(ConcavePolygonShape3D);
+	shape->set_faces(face_points);
+	return shape;
+}
+
+Ref<NavigationMesh> EditorSceneImporterMesh::create_navigation_mesh() {
+	Vector<Face3> faces = get_faces();
+	if (faces.size() == 0) {
+		return Ref<NavigationMesh>();
+	}
+
+	Map<Vector3, int> unique_vertices;
+	LocalVector<int> face_indices;
+
+	for (int i = 0; i < faces.size(); i++) {
+		for (int j = 0; j < 3; j++) {
+			Vector3 v = faces[i].vertex[j];
+			int idx;
+			if (unique_vertices.has(v)) {
+				idx = unique_vertices[v];
+			} else {
+				idx = unique_vertices.size();
+				unique_vertices[v] = idx;
+			}
+			face_indices.push_back(idx);
+		}
+	}
+
+	Vector<Vector3> vertices;
+	vertices.resize(unique_vertices.size());
+	for (Map<Vector3, int>::Element *E = unique_vertices.front(); E; E = E->next()) {
+		vertices.write[E->get()] = E->key();
+	}
+
+	Ref<NavigationMesh> nm;
+	nm.instance();
+	nm->set_vertices(vertices);
+
+	Vector<int> v3;
+	v3.resize(3);
+	for (uint32_t i = 0; i < face_indices.size(); i += 3) {
+		v3.write[0] = face_indices[i + 0];
+		v3.write[1] = face_indices[i + 1];
+		v3.write[2] = face_indices[i + 2];
+		nm->add_polygon(v3);
+	}
+
+	return nm;
+}
+
+extern bool (*array_mesh_lightmap_unwrap_callback)(float p_texel_size, const float *p_vertices, const float *p_normals, int p_vertex_count, const int *p_indices, int p_index_count, float **r_uv, int **r_vertex, int *r_vertex_count, int **r_index, int *r_index_count, int *r_size_hint_x, int *r_size_hint_y, int *&r_cache_data, unsigned int &r_cache_size, bool &r_used_cache);
+
+struct EditorSceneImporterMeshLightmapSurface {
+	Ref<Material> material;
+	LocalVector<SurfaceTool::Vertex> vertices;
+	Mesh::PrimitiveType primitive = Mesh::PrimitiveType::PRIMITIVE_MAX;
+	uint32_t format = 0;
+	String name;
+};
+
+Error EditorSceneImporterMesh::lightmap_unwrap_cached(int *&r_cache_data, unsigned int &r_cache_size, bool &r_used_cache, const Transform &p_base_transform, float p_texel_size) {
+	ERR_FAIL_COND_V(!array_mesh_lightmap_unwrap_callback, ERR_UNCONFIGURED);
+	ERR_FAIL_COND_V_MSG(blend_shapes.size() != 0, ERR_UNAVAILABLE, "Can't unwrap mesh with blend shapes.");
+
+	Vector<float> vertices;
+	Vector<float> normals;
+	Vector<int> indices;
+	Vector<float> uv;
+	Vector<Pair<int, int>> uv_indices;
+
+	Vector<EditorSceneImporterMeshLightmapSurface> lightmap_surfaces;
+
+	// Keep only the scale
+	Transform transform = p_base_transform;
+	transform.origin = Vector3();
+	transform.looking_at(Vector3(1, 0, 0), Vector3(0, 1, 0));
```

Comment:
@reduz This line of code is bogus. `looking_at` returns a copy, so this needs to be

```cpp
transform = transform.looking_at(Vector3(1, 0, 0), Vector3(0, 1, 0));
```

However, why does it point towards +X? If we want to "keep only the scale", we should just not have any rotation, in which case this would be even better:

```cpp
transform.basis = transform.basis.get_scale().to_diagonal_matrix();
```

Reply:
Quoting from my PR description `"WARNING: Lightmap UV unwrap is not working, it needs to be re-made."`
However that said, I just copied this over from Mesh so its likely broken on 3.x, I would talk to @JFonS to see what the intended behavior is.

Reply:
I will go over the UV generation code because there are a bunch of issues, it seems to be working fine for the most part, but some corner cases could cause trouble.

 In this case `to_diagonal_matrix` would be a good fit, I didn't know it existed at the time I wrote this code.

Reply:
I checked the same code in 3.x and both issues (the `looking_at` misuse and the cahce resizing undefied behaviour) are not there. So looks like I fixed those issues but forgot to propagate them to master.